state/remoteapplication.go

```diff
@@ -488,16 +488,50 @@ func (s *RemoteApplication) SetStatus(info status.StatusInfo) error {
 	if !info.Status.KnownWorkloadStatus() {
 		return errors.Errorf("cannot set invalid status %q", info.Status)
 	}
+
+	// If the remote application's been terminated then we aren't
+	// going to see any more state changes for its units - ensure that
+	// they've left scope on all of their relations.
+	var extraOps func() ([]txn.Op, error)
```

Comment:
This I think is the wrong layer to do this in.
Ideally, there'd be a RemoteApplicationTerminated() API call the worker would invoke (which would set status to terminated and do the units thing).
As a compromise here though, to avoid a new facade call, we would move this logic up to the SetRemoteApplicationsStatus() apiserver call. And we'd introduce a ModelOperation implementation to do the status set and unit thing which would be invoked by the apiserver call. Just a few extra lines over what's in the PR now but much cleaner.

Reply:
Checking I understand what you mean: in SetRemoteApplicationsStatus, if the target status is Terminated we'd construct a TerminateRemoteApplicationOperation which would do the status setting and leave scope on the relation units (so they still happen in the same txn).  If the status was anything else we just call remoteApp.SetStatus as normal. Is that right?

Reply:
yup